doc/design.md

```diff
@@ -0,0 +1,128 @@
+
+# Levity remote task runner  
+
+## High-Level Description
+
+`Levity` is a (small part of a) distributed task-running system, consisting of:
+
+1. a server application that can schedule and monitor the execution of arbitrary processes, & an API that exposes that functionality, and
+2. a client application that uses the API to interact with the service.
+
+## Intended Use
+
+`Levity` is intended to be used as an exercise as part of the Gravitational interview process. As this is an exercise, most design tradeoffs will be made in favour of simple implementation over producing an infinitely-robust, battle-hardened production system.
+
+That said, the code that *is* implemented is expected to be robust, clean, modular, and race-free.
+
+## Implementation Language
+
+Go 1.15
+
+## Repo Layout 
+
+The project will use the [Within Go Repo Layout](https://christine.website/blog/within-go-repo-layout-2020-09-07).
+
+## RPC mechanism
+
+`Levity` will use gRPC as its underlying RPC mechanism. The reasons for this are:
+
+* gRPC has built-in support for secure transport and authentication mechanisms, obviating the need to implement these features independently, or wrap them manually around a web service,
+* gRPC mandates the use of a Protocol Buffer description of the service, and having an implementation-independent source of truth about the API is always useful, and
+* I've never used it before, so it seems as good a time to learn as any.
+
+## API
+
+The API exposed by the server is described by the protocol buffer specification [here](../api/levity.proto).
+
+## Task Lifecycle
+
+1. Task is started by a call to `StartTask`.
+2. The client repeatedly polls server for task completion via `QueryTask`.
+3. The client may elect to kill task at any time with `SignalTask`. The
+   server will try to shut it down gracefully at first and then brutally
+   after some (server-specified) timeout.
+4. The client may fetch the logs with `FetchLogs` at any time after
+   starting the task, but will always receive the entire log at the time
+   of the call. There is no mechanism to retreive a partial log.
+5. After some inactivity timeout, the task record on the server is deleted.
+   This implies that the task, if still running, is killed and the logs are no longer retrievable.
+
+## Security Concerns
+
+### Transport
+
+The system will use TLS 1.3 for transport level security. The server will
+require a cetificate issued by a CA trusted by any potential clients.
+
+Providing a system for issuing a certificate and having users install the CA root certificate is beyond the scope of this exercise.
```

Comment:
So how are we going to be able to try it out when it's implemented? I.e. where do we get the CA/keypair for client/server from? Will you provide a script for generating the secrets, or check them into the repo?

Reply:
In order of _de_scending preference:

 1. There will be a makefile target to generate the CA root cert and derive the client and server certs from scratch, or
 2. They will be checked into the repo as independent files, or
 3. They will be baked into the source

Obviously 3 & 4 are not things you would consider doing outside the bounds of an exercise like this.

Reply:
I've added a brief discussion of this to the _Testing Concerns_ section.

Reply:
Okay, I see, I guess it's just the way it's phrased gave an impression they wouldn't be included in any form.

Yep, the options sounds good, it's ok to check secrets into the repo to keep things simple in this case.